File: built-in plugins dsp/mb-sampler-texture.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbSamplerTexturePlugin: PluginDefinition = { id: 'mb-sampler-texture', slug: 'mb-sampler-texture', name: 'MB Texture Sampler', category: 'instrument', type: 'sampler', version: '1.0.0', description: 'Ambient textures and soundscapes', author: 'Max Booster', oscillators: [], envelope: { attack: 1.0, decay: 1.0, sustain: 0.9, release: 2.0 }, parameters: [{ id: 'texture', name: 'Texture', type: 'float', defaultValue: 0.5, minValue: 0, maxValue: 1, automatable: true }, { id: 'space', name: 'Space', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true }, { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.7, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { texture: 0.5, space: 0.7, volume: 0.7 } };

export default MbSamplerTexturePlugin;

File: built-in plugins dsp/mb-sampler-texture.cpp
/**
 * MB Texture Sampler
 * Category : instrument
 * Type     : sampler
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Ambient textures and soundscapes
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_SAMPLER_TEXTURE_H
#define MB_SAMPLER_TEXTURE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbSamplerTexture : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-sampler-texture";
    static constexpr const char* PLUGIN_NAME    = "MB Texture Sampler";
    static constexpr const char* PLUGIN_TYPE    = "sampler";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float texture = 0.5f;  // range [0, 1]
    float space = 0.7f;  // range [0, 1]
    float volume = 0.7f;  // range [0, 1]
    };

    MbSamplerTexture() = default;
    ~MbSamplerTexture() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.texture = std::clamp(params.texture, 0f, 1f);
        params.space = std::clamp(params.space, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Texture Sampler
        return input;
    }
};

#endif // MB_SAMPLER_TEXTURE_H
